### PR TITLE
SAC Performance Fixes

### LINKED ISSUE #5974: SAC performance on mujoco
Hi,
I find it difficult to figure out working parameters for SAC. Is there just some standard examples like e.g. in the original softlearning environment?

### PR DESCRIPTION
**Pull request by michaelzhiluo**
## Why are these changes needed?

SAC's sample efficiency on RLLib is now comparable to the original SoftLearning repo.
There were three bugs that we observed:
1) The target Q network's entropy term should be - alpha * pi(a|s_(t+1))
2) The actor policy loss should be over the minimum of Q values.
3) There should be three-four optimizers for actor, critic, and alpha variables respectively. Added `apply_gradients_fn` in `sac_policy.py` to fix the bug of only one optimizer applying gradients.

## Checks

- [x ] I've run `scripts/format.sh` to lint the changes in this PR.
Closes https://github.com/ray-project/ray/issues/5974
Closes

### CODE REVIEW

File: rllib/agents/sac/sac_policy.py [hunk 1]
```diff
@@ -111,10 +111,13 @@ def unsquash_actions(actions):
 
     squashed_stochastic_actions, log_pis = policy.model.get_policy_output(
         model_out, deterministic=False)
-    stochastic_actions = unsquash_actions(squashed_stochastic_actions)
+    stochastic_actions = squashed_stochastic_actions if config[
+        "normalize_actions"] else unsquash_actions(squashed_stochastic_actions)
```
Comment (ericl) [inline rllib/agents/sac/sac_policy.py]: Which one should we use? Do either work?

File: rllib/agents/sac/sac_model.py [hunk 2]
```diff
@@ -111,17 +110,68 @@ def __init__(self,
         shift_and_log_scale_diag = tf.keras.layers.Concatenate(axis=-1)(
             [shift, log_scale_diag])
 
-        raw_action_distribution = tfp.layers.MultivariateNormalTriL(
-            self.action_dim)(shift_and_log_scale_diag)
+        batch_size = tf.keras.layers.Lambda(lambda x: tf.shape(input=x)[0])(
```
Comment (ericl) [inline rllib/agents/sac/sac_model.py]: Hm what are the changes here for?

File: rllib/tuned_examples/pendulum-sac.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Pendulum SAC can attain -150+ reward in 5-6k timesteps (*disregarding initial exploration steps like in softlearning codebase)
+# Configurations are the similar to original softlearning/sac codebase
```
Comment (ericl) [inline rllib/tuned_examples/pendulum-sac.yaml]: Can you update the pendulum-sac regression test file? (assuming it is reliable enough for that?)

**Comment by ericl** [comment]
Oops, re-opened. Could you also add some graphs for HalfCheetah or other mujoco envs? Should be fine to use the hyperparams from softlearning.

File: rllib/agents/sac/sac_policy.py (hunk 1 above)
Comment (michaelzhiluo) [reply rllib/agents/sac/sac_policy.py]: The normalize actions wrapper works better empirically.

File: rllib/agents/sac/sac_model.py (hunk 2 above)
Comment (michaelzhiluo) [reply rllib/agents/sac/sac_model.py]: Replacing distributional computation with more explicit code from softlearning codebase.

File: rllib/agents/trainer.py [hunk 4]
```diff
@@ -493,7 +496,12 @@ def _setup(self, config):
                 self.env_creator = _global_registry.get(ENV_CREATOR, env)
             else:
                 import gym  # soft dependency
-                self.env_creator = lambda env_config: gym.make(env)
+                env_ = gym.make(env)
+                act_wrap = NormalizeActionWrapper
+                if config["normalize_actions"]:
+                    env_ = act_wrap(env_.env)
+                self.env_creator = lambda env_config: env_
```
Comment (ericl) [inline rllib/agents/trainer.py]: Can you make it so this wrapper works for both the gym and non-gym cases?

File: rllib/agents/sac/sac_policy.py (hunk 1 above)
Comment (ericl) [reply rllib/agents/sac/sac_policy.py]: Ok, let's remove the squashing then and have sac.py enable normalize_actions by default.

File: rllib/agents/sac/sac_model.py (hunk 2 above)
Comment (hartikainen) [reply rllib/agents/sac/sac_model.py]: My policy implementation in softlearning is not ideal and could be refactored to be a lot cleaner, which is why I think we shouldn't bring those changes over here. Instead, it would be better to refactor both of them, but that can be left for future PR.

File: rllib/agents/sac/sac_policy.py [hunk 5]
```diff
@@ -165,9 +172,8 @@ def actor_critic_loss(policy, model, _, train_batch):
     q_t_selected = tf.squeeze(q_t, axis=len(q_t.shape) - 1)
     if policy.config["twin_q"]:
         twin_q_t_selected = tf.squeeze(twin_q_t, axis=len(q_t.shape) - 1)
-        q_tp1 = tf.minimum(q_tp1, twin_q_tp1)
-
-    q_tp1 -= tf.expand_dims(alpha * log_pis_t, 1)
+        q_tp1 = tf.reduce_min((q_tp1, twin_q_tp1), axis=0)
+    q_tp1 -= alpha * log_pis_tp1
```
Comment (hartikainen) [inline rllib/agents/sac/sac_policy.py]: Oh, nice catch!

File: rllib/agents/sac/sac_policy.py [hunk 6]
```diff
@@ -165,9 +172,8 @@ def actor_critic_loss(policy, model, _, train_batch):
     q_t_selected = tf.squeeze(q_t, axis=len(q_t.shape) - 1)
     if policy.config["twin_q"]:
         twin_q_t_selected = tf.squeeze(twin_q_t, axis=len(q_t.shape) - 1)
-        q_tp1 = tf.minimum(q_tp1, twin_q_tp1)
-
-    q_tp1 -= tf.expand_dims(alpha * log_pis_t, 1)
+        q_tp1 = tf.reduce_min((q_tp1, twin_q_tp1), axis=0)
```
Comment (hartikainen) [inline rllib/agents/sac/sac_policy.py]: Just to make sure: both of these result in the same `q_tp1`? I think this part of code will fail silently if one of the dimensions end up being `(N, )` instead of `(N, 1)`.

File: rllib/agents/sac/sac_policy.py [hunk 7]
```diff
@@ -249,6 +287,29 @@ def gradients(policy, optimizer, loss):
     return grads_and_vars
 
 
+def apply_gradients(policy, optimizer, grads_and_vars):
+    actor_apply_ops = policy._actor_optimizer.apply_gradients(
+        policy._actor_grads_and_vars,
+        global_step=tf.train.get_or_create_global_step())
```
Comment (hartikainen) [inline rllib/agents/sac/sac_policy.py]: Does `apply_gradients` automatically increment the global step? If it does, then I think we should not have this both here and below.

File: rllib/env/normalize_actions.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import gym
```
Comment (hartikainen) [inline rllib/env/normalize_actions.py]: This wrapper has recently been added in the gym master. Not sure if we're able upgrade gym version for rllib, but if we are, I think we should use the standard implementation.

https://github.com/openai/gym/blob/master/gym/wrappers/rescale_action.py

File: rllib/tuned_examples/pendulum-sac.yaml [hunk 9]
```diff
@@ -0,0 +1,37 @@
+# Pendulum SAC can attain -150+ reward in 5-6k timesteps (*disregarding initial exploration steps like in softlearning codebase)
```
Comment (hartikainen) [inline rllib/tuned_examples/pendulum-sac.yaml]: I think it's generally better to include the exploration steps in the sample count, i.e. just add 1k to those numbers.

**Review by hartikainen — APPROVED** [review]
Thanks a lot for fixing this @michaelzhiluo!

I'm curious, what is the change that actually made the previous implementation not match softlearning's performance. I'd guess it's just the bug in the Q-target computation, but would be interesting to hear if e.g. having independent optimizers makes difference here. I wouldn't be surprised at all if it does.

File: rllib/env/normalize_actions.py (hunk 8 above)
Comment (ericl) [reply rllib/env/normalize_actions.py]: I think that's fine yeah, we can raise an error telling the user to upgrade if it's not found.

File: rllib/agents/sac/sac_policy.py (hunk 1 above)
Comment (hartikainen) [reply rllib/agents/sac/sac_policy.py]: Just to make sure, even with the action normalization/rescaling, we still need to squash the actions in SAC. What we don't have to do is unsquash them on this level.

Ideally, I think SAC should be able to just handle different action ranges though and there's no reason why it couldn't do it. In the softlearning implementation, I just haven't bothered to change things because we pretty much automatically scale all the action spaces to be between `-1...1`.

File: rllib/agents/trainer.py (hunk 4 above)
Comment (michaelzhiluo) [reply rllib/agents/trainer.py]: `self.env_creator` is a lambda function. When the env is not from gym, not sure on how to modify the environment before getting the env_config parameter.

File: rllib/agents/sac/sac_policy.py (hunk 7 above)
Comment (michaelzhiluo) [reply rllib/agents/sac/sac_policy.py]: Fixed! Only one optimizer (alpha optimizer) will increment the global step.

File: rllib/tuned_examples/pendulum-sac.yaml (hunk 9 above)
Comment (michaelzhiluo) [reply rllib/tuned_examples/pendulum-sac.yaml]: Fixed!

File: rllib/tuned_examples/pendulum-sac.yaml (hunk 3 above)
Comment (michaelzhiluo) [reply rllib/tuned_examples/pendulum-sac.yaml]: Done!

File: rllib/agents/sac/sac_model.py (hunk 2 above)
Comment (michaelzhiluo) [reply rllib/agents/sac/sac_model.py]: ^^ The distributional code was a lot more concise. I switched back to softlearning database to debug via making everything deterministic.

File: rllib/agents/sac/sac_policy.py (hunk 6 above)
Comment (michaelzhiluo) [reply rllib/agents/sac/sac_policy.py]: Just checked on pdb and both q_tp1 and twin_q_tp1 are (N,1). The policy graphs for softlearning and Rllib SAC are pretty much identical. Debugged SAC by setting random seeds to constant values and made everything deterministic. Softlearning and Rllib SAC output the same loss function values even after multiple gradient steps...so I believe any other deltas could be from code outside of SAC.

Comment (hartikainen) [reply rllib/agents/sac/sac_policy.py]: Nice, sounds good!

File: rllib/agents/trainer.py (hunk 4 above)
Comment (ericl) [reply rllib/agents/trainer.py]: Why not

```
self.env_creator = ...  # anything

if config["normalize_actions"]:
   old_creator = self.env_creator

   def make_normalized(env_config):
       env = old_creator(env_config)
       assert isinstance(env, gym.Env)
       env = NormalizeActionWrapper(env)
       return env

   self.env_creator = make_normalized
```

**Comment by ericl** [comment]
Jenkins retest this please

**Comment by ericl** [comment]
Jenkins retest this please

File: rllib/agents/sac/sac_model.py [hunk 10]
```diff
@@ -111,17 +110,68 @@ def __init__(self,
...
+            lambda batch_size: base_distribution.sample(batch_size))(
+                batch_size)
+
+        self.shift_and_log_scale_diag = latents
+        self.latents_model = tf.keras.Model(self.model_out, latents)
+
+        def raw_actions_fn(inputs):
+            shift, log_scale_diag, latents = inputs
+            bijector = tfp.bijectors.Affine(
+                shift=shift, scale_diag=tf.exp(log_scale_diag))
```
Comment (stepjam) [inline rllib/agents/sac/sac_model.py]: I seem to be getting an dimension incompatibility since this change when running SAC on any environment. I have tracked it down to here. Any idea @michaelzhiluo ?

Comment (michaelzhiluo) [reply rllib/agents/sac/sac_model.py]: Not sure, what are your Tensorflow dependencies?

Comment (michaelzhiluo) [reply rllib/agents/sac/sac_model.py]: Fixed here: https://github.com/ray-project/ray/pull/6642

**Comment by kapsl** [comment]
I tested the new implementation of SAC and it is indeed now closer to the original implementation regarding performance. Still I got the problem, that it is close to the performance although I'm using 27 workers compared to 1 worker in the old implementation. Does the parallelization work at all for SAC? In the code it is written:

# Number of workers for collecting samples with. This only makes sense
    # to increase if your environment is particularly slow to sample, or if
    # you"re using the Async or Ape-X optimizers.
    "num_workers": 0,

How could i set these optimizers? Or is this a copy and paste mistake? Till now it does not really make sense to use Ray if I get a little bit lower performance with 27 workers.

**Comment by ericl** [comment]
@kapsl you'll only get a speedup if your simulator is slow. If you're using a fast simulator such as mujoco, then the learning will be the bottleneck and adding rollout workers won't really help.

You could try making an Apex-SAC variant (though afaik no one has tried this), see `dqn/apex.py` for an example on how to set this up.

**Comment by kapsl** [comment]
you mean like having more gradient steps per optimization step?! and learning rate?

File: rllib/agents/sac/sac_policy.py [hunk 11]
```diff
@@ -155,6 +158,10 @@ def actor_critic_loss(policy, model, _, train_batch):
 
     # Q-values for current policy (no noise) in given current state
     q_t_det_policy = model.get_q_values(model_out_t, policy_t)
+    if policy.config["twin_q"]:
+        twin_q_t_det_policy = model.get_q_values(model_out_t, policy_t)
```
Comment (mattearllongshot) [inline rllib/agents/sac/sac_policy.py]: Should this be using `model.get_twin_q_values`?  It seems the `min` on the next line would just take the minimum of identical values.

Comment (michaelzhiluo) [reply rllib/agents/sac/sac_policy.py]: You are right... could you see how changing this line of code does for `Pendulum-v0` in tuned examples? Hopefully it should improve sample efficiency.

Comment (ericl) [reply rllib/agents/sac/sac_policy.py]: FYI @sven1977

Comment (mattearllongshot) [reply rllib/agents/sac/sac_policy.py]: image
blue = with fix, orange = without fix.  It doesn't seem to have made a huge difference, although perhaps we wouldn't expect it to on such a small task?  I would give HalfCheetah-v3 a go, however I do not have a MuJuCo license.

Comment (michaelzhiluo) [reply rllib/agents/sac/sac_policy.py]: Send me a message at michael.luo@berkeley.edu. I'll be happy to help you :).

Comment (hartikainen) [reply rllib/agents/sac/sac_policy.py]: I think `Pendulum-v0` is probably too trivial environment for the min-Q trick to actually have an effect. If you compare this in environment like `Ant-v3` or `Humanoid-v3`, I'd expect both the final performance and also the learning speed to be much better with min-Q.

Comment (mattearllongshot) [reply rllib/agents/sac/sac_policy.py]: image
Here are the results for HalfCheetah-v3 using halfcheetah-sac.yaml from the repository.  I checked out the 0.8.2 tag as master had stability issues (`raise MujocoException(warn + 'Check for NaN in simulation.')`), and commented out the `exploration_enabled` line in the YAML file as I was getting `Unknown config parameter `exploration_enabled`` errors.

The grey line is WITH the fix, green line WITHOUT.  So a bit of a counter intuitive result.  Could it be that because it was hyper parameter tuned without the fix, changing the algorithm in this way has in fact caused degradation?  If so we'd want to re-tune with the fix to do a like for like comparison (unfortunately I don't have resources for this at the moment).

Comment (hartikainen) [reply rllib/agents/sac/sac_policy.py]: `raise MujocoException(warn + 'Check for NaN in simulation.'` in case of SAC is very likely due to numerically unstable implementation of the squashed head (`SquashedGaussian` here if I remember seeing correctly) of the policy. See e.g. https://github.com/tensorflow/probability/issues/318.

Comment (hartikainen) [reply rllib/agents/sac/sac_policy.py]: Ignore what I said about `SquashedGaussian`, since this PR is still using the `SquashBijector`. That should be okay.

Comment (mattearllongshot) [reply rllib/agents/sac/sac_policy.py]: Actually, I was using `SquashedGaussian` (`master` latest)  when I saw the NaN error, and it disappeared when I downgraded.  I commented on this PR because it was where the `get_q_values` line in question was introduced.

Comment (ericl) [reply rllib/agents/sac/sac_policy.py]: @sven1977

Comment (ericl) [reply rllib/agents/sac/sac_policy.py]: Btw @mattearllongshot could you report your TF/TFP version and how to reproduce? The reason is we didn't see the instability in our testing.

Comment (sven1977) [reply rllib/agents/sac/sac_policy.py]: @mattearllongshot I'll take another look at the SquashedGaussian code to see, whether I can break it as well. Haven't seen any issues when running it against HalfCheetah a couple of times last WE (actually performed slightly better than our previous baseline (~14k vs 13k)).

Comment (sven1977) [reply rllib/agents/sac/sac_policy.py]: @mattearllongshot And thanks for catching that bug. :)

Comment (hartikainen) [reply rllib/agents/sac/sac_policy.py]: Here's a related tensorflow-probability issue if that helps: https://github.com/tensorflow/probability/issues/318.

Comment (mattearllongshot) [reply rllib/agents/sac/sac_policy.py]: I'm using

```
tensorflow==2.1.0
tensorflow-estimator==2.1.0
tensorflow-probability==0.9.0
```

on commit 9964657815fdd125d364ef8a09bd3cd046399611, running `rllib train -f ./rllib/tuned_examples/halfcheetah-sac.yaml`.  On this occasion I saw the `Check for NaN in simulation.` error after 32 iterations

Comment (mattearllongshot) [reply rllib/agents/sac/sac_policy.py]: Shall I raise a bug for these two so they don't get lost?

Comment (mattearllongshot) [reply rllib/agents/sac/sac_policy.py]: or rather, two bugs
